File: interview/LeetCode/DecemberLeetCodingChallenge/Week1/Shortest_Word_Distance.cpp
// Copyright (c) hoduchieu01
/*
    Given a list of words and two words word1 and word2, return the shortest distance between these two words in the list.
    Algorithm: Brute Force
    Time Complexity: O(n*n) & Space Complexity: O(1)
    // strcmp(string 1, string 2) == 0 when 2 strings equal
*/
int min(int a, int b){
    if(a > b) 
        return b;
    else
        return a;
}

int shortestDistance(char ** words, int wordsSize, char * word1, char * word2){
    int minDistance = wordsSize;
    for(int i = 0; i < wordsSize; i++){
        if(strcmp(words[i], word1) == 0){
            for(int j = 0; j < wordsSize; j++){
                if(strcmp(words[j], word2) == 0){
                    minDistance = min(minDistance, abs(i-j));
                }
            }
        }
    }
    return minDistance;
}

File: interview/LeetCode/DecemberLeetCodingChallenge/Week1/Shortest_Word_Distance.py
# Copyright (c) hoduchieu01
class Solution:
    def shortestDistance(self, words: List[str], word1: str, word2: str) -> int:
        word_dict = defaultdict(list)
        for i, w in enumerate(words):
            if w in (word1, word2): 
                word_dict[w].append(i)
        minDistance = float('inf')
        for i in word_dict[word1]:
            for j in word_dict[word2]:
                minDistance = min(minDistance, abs(i-j))
        return minDistance